### DISCUSSION TITLE
Make state changes persist outside of function call

### DISCUSSION THREAD

**Discussion by mfdorst** [nushell/nushell#10069] [Q&A]
The idea of an immutable language is pretty awesome, however, it does raise some usability concerns. There are a few things I have been doing in my other shells that are harder (or impossible?) in nushell.

```
if (which exa | length) > 0 {
  alias l = exa --git -l
  alias la = exa --git -la
}
else {
  alias l = ls
  alias la = ls -a
}
```
The aliases do not persist outside of the `if` statement, so you have to instead do
```
alias l = if (which exa | length) > 0 { exa --git -l } else { ls }
alias la = if (which exa | length) > 0 { exa --git -la } else { ls -a }
```
which is fine, but unwieldy when there are lots of aliases to set conditionally.

More recently I've run into one that I'm not sure how to solve.
```
def mcd [dir] {
  mkdir $dir
  cd $dir
}
```
The `cd` does not persist after the function call. Is there a workaround for this? And more broadly, might it make sense to have a way to allow global state changes to escape a function call?

**Comment by ghost** [comment]
> The aliases do not persist outside of the if statement, so you have to instead do

afaik thats a issue with the language.  
Solutions im aware of:
- `alias foo = if … {a} else {b}`
- for config: generate a module script (`help module`) in `env.nu` and load it in `config.nu` (you cant generate and load in the same file)
- completly avoid it (device specific config, or termial specific config (launch args), …)

> The cd does not persist after the function call. Is there a workaround for this?

environmental variables (including PWD, which describes your current directory) are function scoped per default - use `def-env` instead of `def` to use the global env vars instead

**Comment by kubouch** [comment]
When https://github.com/nushell/nushell/pull/9499 is merged after the next release, you'll be able to use parse-time evaluation to conditionally import a module depending on a condition, for example:
```nushell
const file_to_import = if (...condition...) { 'exa_aliases.nu' } else { 'orig_aliases.nu' }
use $file_to_import *
```

This will have the limitation that the condition would need to be a parse-time constant, and `which` command likely won't be possible to evaluate at parse time (we're not yet 100% how far we can go with it as we just started experimenting with it).
